src/infra/core/src/remote_alias_resolver_impl.rs

```diff
@@ -0,0 +1,308 @@
+                }
+            }
+        } else {
+            if let Some(remote_alias) = self
+                .fetch_remote_alias(local_dataset_handle, remote_alias_kind)
+                .await?
+            {
+                return Ok(remote_alias);
+            }
+            let remote_repo_names: Vec<_> = self.remote_repo_reg.get_all_repositories().collect();
+            if remote_repo_names.len() > 1 {
+                return Err(ResolveAliasError::AmbiguousRepository);
+            }
+            if let Some(repository_name) = remote_repo_names.first() {
+                repo_name = repository_name.clone();
+            } else {
+                return Err(ResolveAliasError::EmptyRepositoryList);
+            }
+        }
+        let remote_repo = self.remote_repo_reg.get_repository(&repo_name).int_err()?;
+
+        if account_name.is_none() {
+            let access_token_maybe = self
+                .access_token_resolver
+                .resolve_odf_dataset_access_token(&remote_repo.url);
+            account_name = RemoteAliasResolverApiHelper::resolve_remote_account_name(
+                &remote_repo.url,
+                access_token_maybe,
+            )
+            .await
+            .int_err()?;
+        }
+        let push_dataset_name = dataset_name.unwrap_or(
+            self.resolve_remote_dataset_name(local_dataset_handle, &remote_repo.url)
+                .await?,
+        );
+
+        let remote_alias =
+            self.combine_remote_alias(&remote_repo.url, account_name, &push_dataset_name)?;
+
+        return Ok(remote_alias);
+    }
+
+    // TODO: avoid traversing all datasets for every alias
+    async fn inverse_lookup_dataset_by_alias(
+        &self,
+        transfer_ref: &odf::TransferDatasetRef,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<odf::DatasetHandle, ResolveAliasError> {
+        // Do a quick check when remote and local names match
+        if let odf::TransferDatasetRef::RemoteRef(remote_ref) = transfer_ref {
+            if let Some(remote_name) = remote_ref.dataset_name()
+                && let Some(local_handle) = self
+                    .dataset_repo
+                    .try_resolve_dataset_ref(
+                        &odf::DatasetAlias::new(None, remote_name.clone()).as_local_ref(),
+                    )
+                    .await?
+                && self
+                    .remote_alias_reg
+                    .get_remote_aliases(&local_handle.as_local_ref())
+                    .await
+                    .int_err()?
+                    .contains(remote_ref, remote_alias_kind)
+            {
+                return Ok(local_handle);
+            }
+
+            // No luck - now have to search through aliases
+            use tokio_stream::StreamExt;
+            let mut datasets = self.dataset_repo.get_all_datasets();
+            while let Some(dataset_handle) = datasets.next().await {
+                let dataset_handle = dataset_handle?;
```

Comment:
Let's make sure we have a test for this state. I'm going to refactor this soon to list the handles from the database entries, but I'm not sure what exactly reproduces this state.

Reply:
It is tasted in `test_push` scope